pkg/kubelet/server/stats/summary.go

```diff
@@ -0,0 +1,354 @@
+		Recursive: true,
+	}
+	infos, err := sp.provider.GetContainerInfoV2("/", options)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := sp.provider.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	nodeConfig := sp.provider.GetNodeConfig()
+	rootFsInfo, err := sp.provider.RootFsInfo()
+	if err != nil {
+		return nil, err
+	}
+	imageFsInfo, err := sp.provider.DockerImagesFsInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	sb := &summaryBuilder{node, nodeConfig, rootFsInfo, imageFsInfo, infos}
+	return sb.build()
+}
+
+// summaryBuilder aggregates the datastructures provided by cadvisor into a Summary result
+type summaryBuilder struct {
+	node        *api.Node
+	nodeConfig  cm.NodeConfig
+	rootFsInfo  cadvisorapiv2.FsInfo
+	imageFsInfo cadvisorapiv2.FsInfo
+	infos       map[string]cadvisorapiv2.ContainerInfo
+}
+
+// build returns a Summary from aggregating the input data
+func (sb *summaryBuilder) build() (*Summary, error) {
+	rootInfo, found := sb.infos["/"]
+	if !found {
+		return nil, fmt.Errorf("Missing stats for root container")
+	}
+	cstat, found := sb.latestContainerStats(&rootInfo)
+	if !found {
+		return nil, fmt.Errorf("Missing stats for root container")
+	}
+
+	rootStats := sb.containerInfoV2ToStats("", &rootInfo)
+	nodeStats := NodeStats{
+		NodeName: sb.node.Name,
+		CPU:      rootStats.CPU,
+		Memory:   rootStats.Memory,
+		Network:  sb.containerInfoV2ToNetworkStats(&rootInfo),
+		Fs:       sb.getNodeFsStats(),
+	}
+
+	systemContainers := map[string]string{
+		SystemContainerKubelet: sb.nodeConfig.KubeletContainerName,
+		SystemContainerRuntime: sb.nodeConfig.DockerDaemonContainerName, // TODO: add support for other runtimes
+		SystemContainerMisc:    sb.nodeConfig.SystemContainerName,
+	}
+	for sys, name := range systemContainers {
+		if info, ok := sb.infos[name]; ok {
+			nodeStats.SystemContainers = append(nodeStats.SystemContainers, sb.containerInfoV2ToStats(sys, &info))
+		}
+	}
+
+	summary := Summary{
+		Time: unversioned.NewTime(cstat.Timestamp),
+		Node: nodeStats,
+		Pods: sb.buildSummaryPods(),
+	}
+	return &summary, nil
+}
+
+// getNodeFsStats returns the node root fs stats
+func (sb *summaryBuilder) getNodeFsStats() *FsStats {
+	return &FsStats{
+		AvailableBytes: &sb.rootFsInfo.Available,
+		CapacityBytes:  &sb.rootFsInfo.Capacity,
+		UsedBytes:      &sb.rootFsInfo.Usage,
+	}
+}
+
+// containerInfoV2FsStats populates the container fs stats
+func (sb *summaryBuilder) containerInfoV2FsStats(
+	info *cadvisorapiv2.ContainerInfo,
+	cs *ContainerStats) {
+
+	// The container logs live on the node rootfs device
+	cs.Logs = &FsStats{
+		AvailableBytes: &sb.rootFsInfo.Available,
+		CapacityBytes:  &sb.rootFsInfo.Capacity,
+	}
+
+	// The container rootFs lives on the imageFs devices (which may not be the node root fs)
+	cs.Rootfs = &FsStats{
+		AvailableBytes: &sb.imageFsInfo.Available,
+		CapacityBytes:  &sb.imageFsInfo.Capacity,
+	}
+
+	lcs, found := sb.latestContainerStats(info)
+
```

Comment:
(tiny nit) I prefer `if !found | !ok | err != nil | etc. {` type checks to be right next to where the variable is set (i.e. remove this blank line)